src/ds/ring_buffer.h

```diff
@@ -144,8 +145,9 @@ namespace ringbuffer
   {
     static inline uint64_t read64_impl(const BufferDef& bd, size_t index)
     {
-      uint64_t r = *reinterpret_cast<volatile uint64_t*>(bd.data + index);
-      atomic_thread_fence(std::memory_order_acq_rel);
+      uint64_t r = 0;
+      __atomic_load(
+        reinterpret_cast<uint64_t*>(bd.data + index), &r, __ATOMIC_ACQUIRE);
```

Comment:
`volatile` is a keyword for preventing optimization with special memory not for allowing concurrency access. So for the old code, TSAN shows warnings like:

```
==================
WARNING: ThreadSanitizer: data race (pid=16999)
  Read of size 8 at 0x7b080011a530 by main thread:
    #0 ringbuffer::(anonymous namespace)::read64_impl(ringbuffer::BufferDef const&, unsigned long) /home/takurosato/CCF/build/../src/ds/test/../ring_buffer.h:148:20 (ds_test+0x1cfa10) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #1 ringbuffer::Reader::read64(unsigned long) /home/takurosato/CCF/build/../src/ds/test/../ring_buffer.h:175:14 (ds_test+0x2342f8) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #2 ringbuffer::Reader::read(unsigned long, std::__1::function<void (unsigned int, unsigned char const*, unsigned long)>) /home/takurosato/CCF/build/../src/ds/test/../ring_buffer.h:210:23 (ds_test+0x232938) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #3 DOCTEST_ANON_FUNC_76() /home/takurosato/CCF/build/../src/ds/test/ring_buffer.cpp:612:27 (ds_test+0x1cc68f) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #4 doctest::Context::run() /home/takurosato/CCF/build/../3rdparty/test/doctest/doctest.h:7007:21 (ds_test+0x1ba05e) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #5 main /home/takurosato/CCF/build/../3rdparty/test/doctest/doctest.h:7085:71 (ds_test+0x1bb987) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)

  Previous write of size 8 at 0x7b080011a530 by thread T17:
    #0 ringbuffer::Writer::write64(unsigned long, unsigned long) /home/takurosato/CCF/build/../src/ds/test/../ring_buffer.h:432:62 (ds_test+0x278292) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #1 ringbuffer::Writer::prepare(unsigned int, unsigned long, bool, unsigned long*) /home/takurosato/CCF/build/../src/ds/test/../ring_buffer.h:360:7 (ds_test+0x277c9b) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #2 bool ringbuffer::AbstractWriter::write_multiple<serializer::CommonSerializer, unsigned char&>(unsigned int, bool, unsigned char&) /home/takurosato/CCF/build/../src/ds/test/../ring_buffer_types.h:134:35 (ds_test+0x280f7a) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #3 void ringbuffer::AbstractWriter::write_with<serializer::CommonSerializer, unsigned char&>(unsigned int, unsigned char&) /home/takurosato/CCF/build/../src/ds/test/../ring_buffer_types.h:71:7 (ds_test+0x2a4696) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #4 void ringbuffer::AbstractWriter::write<unsigned char&>(unsigned int, unsigned char&) /home/takurosato/CCF/build/../src/ds/test/../ring_buffer_types.h:85:7 (ds_test+0x2a4621) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #5 DOCTEST_ANON_FUNC_76()::$_19::operator()() const /home/takurosato/CCF/build/../src/ds/test/ring_buffer.cpp:603:13 (ds_test+0x1df9d7) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #6 decltype(std::declval<DOCTEST_ANON_FUNC_76()::$_19>()()) std::__1::__invoke[abi:v15007]<DOCTEST_ANON_FUNC_76()::$_19>(DOCTEST_ANON_FUNC_76()::$_19&&) /usr/lib/llvm-15/bin/../include/c++/v1/__functional/invoke.h:394:23 (ds_test+0x1df8b5) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #7 void std::__1::__thread_execute[abi:v15007]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, DOCTEST_ANON_FUNC_76()::$_19>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, DOCTEST_ANON_FUNC_76()::$_19>&, std::__1::__tuple_indices<>) /usr/lib/llvm-15/bin/../include/c++/v1/thread:284:5 (ds_test+0x1df86d) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #8 void* std::__1::__thread_proxy[abi:v15007]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, DOCTEST_ANON_FUNC_76()::$_19>>(void*) /usr/lib/llvm-15/bin/../include/c++/v1/thread:295:5 (ds_test+0x1df4df) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)

  Location is heap block of size 32 at 0x7b080011a520 allocated by main thread:
    #0 operator new(unsigned long) <null> (ds_test+0x1a7546) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #1 void* std::__1::__libcpp_operator_new[abi:v15007]<unsigned long>(unsigned long) /usr/lib/llvm-15/bin/../include/c++/v1/new:246:10 (ds_test+0x207825) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #2 std::__1::__libcpp_allocate[abi:v15007](unsigned long, unsigned long) /usr/lib/llvm-15/bin/../include/c++/v1/new:272:10 (ds_test+0x207762) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #3 std::__1::allocator<unsigned char>::allocate[abi:v15007](unsigned long) /usr/lib/llvm-15/bin/../include/c++/v1/__memory/allocator.h:112:38 (ds_test+0x230591) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #4 std::__1::__allocation_result<std::__1::allocator_traits<std::__1::allocator<unsigned char>>::pointer> std::__1::__allocate_at_least[abi:v15007]<std::__1::allocator<unsigned char>>(std::__1::allocator<unsigned char>&, unsigned long) /usr/lib/llvm-15/bin/../include/c++/v1/__memory/allocate_at_least.h:54:19 (ds_test+0x2303ad) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #5 std::__1::vector<unsigned char, std::__1::allocator<unsigned char>>::__vallocate[abi:v15007](unsigned long) /usr/lib/llvm-15/bin/../include/c++/v1/vector:682:29 (ds_test+0x233c13) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #6 std::__1::vector<unsigned char, std::__1::allocator<unsigned char>>::vector(unsigned long, unsigned char const&) /usr/lib/llvm-15/bin/../include/c++/v1/vector:1108:9 (ds_test+0x233a22) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #7 ringbuffer::TestBuffer::TestBuffer(unsigned long) /home/takurosato/CCF/build/../src/ds/test/../ring_buffer.h:588:31 (ds_test+0x2337f7) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #8 std::__1::__unique_if<ringbuffer::TestBuffer>::__unique_single std::__1::make_unique[abi:v15007]<ringbuffer::TestBuffer, unsigned long const&>(unsigned long const&) /usr/lib/llvm-15/bin/../include/c++/v1/__memory/unique_ptr.h:714:32 (ds_test+0x285361) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #9 DOCTEST_ANON_FUNC_76() /home/takurosato/CCF/build/../src/ds/test/ring_buffer.cpp:586:19 (ds_test+0x1cc1d8) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #10 doctest::Context::run() /home/takurosato/CCF/build/../3rdparty/test/doctest/doctest.h:7007:21 (ds_test+0x1ba05e) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #11 main /home/takurosato/CCF/build/../3rdparty/test/doctest/doctest.h:7085:71 (ds_test+0x1bb987) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)

  Thread T17 (tid=17289, finished) created by main thread at:
    #0 pthread_create <null> (ds_test+0x127bfd) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #1 std::__1::__libcpp_thread_create[abi:v15007](unsigned long*, void* (*)(void*), void*) /usr/lib/llvm-15/bin/../include/c++/v1/__threading_support:376:10 (ds_test+0x2a4069) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #2 std::__1::thread::thread<DOCTEST_ANON_FUNC_76()::$_19, void>(DOCTEST_ANON_FUNC_76()::$_19&&) /usr/lib/llvm-15/bin/../include/c++/v1/thread:311:16 (ds_test+0x1df26f) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #3 DOCTEST_ANON_FUNC_76() /home/takurosato/CCF/build/../src/ds/test/ring_buffer.cpp:597:32 (ds_test+0x1cc345) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #4 doctest::Context::run() /home/takurosato/CCF/build/../3rdparty/test/doctest/doctest.h:7007:21 (ds_test+0x1ba05e) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)
    #5 main /home/takurosato/CCF/build/../3rdparty/test/doctest/doctest.h:7085:71 (ds_test+0x1bb987) (BuildId: 2c47a40c12e3b309b73007356d02220fdd007059)

SUMMARY: ThreadSanitizer: data race /home/takurosato/CCF/build/../src/ds/test/../ring_buffer.h:148:20 in ringbuffer::(anonymous namespace)::read64_impl(ringbuffer::BufferDef const&, unsigned long)
==================
```

For `atomic_thread_fence(std::memory_order_acq_rel);` in the old code, I am not sure the reason that it was there.
[atomic_thread_fence](https://en.cppreference.com/w/cpp/atomic/atomic_thread_fence) should be used with an atomic variable, but it's not clear which atomic variable is the one.
Considering that `read64_impl` does only the fence and the read against volatile pointer, the fence might be intended to be used with the volatile read, but I think it's not supported according to the document.

Since TSAN doesn't complain about this function anymore I think we can confidently remove the fence.